File: Python/Program/Tabel.py
'''
Saya Nabil Hanif Achmaddiredja mengerjakan Latihan Praktikum 4 dalam 
mata kuliah Desain dan Pemrograman Berorientasi Objek untuk 
keberkahanNya maka saya tidak melakukan kecurangan seperti yang telah dispesifikasikan. Aamiin.
'''

class Tabel:
    __baris = 0
    __kolom = 0
    
    # Constructor
    def __init__(self, baris=0, kolom=0):
        self.__baris = baris
        self.__kolom = kolom

    # Getter and Setter
    # Getter 
    def getBaris(self):
        return self.__baris

    def getKolom(self):
        return self.__kolom

    # Setter 
    def setBaris(self, baris):
        self.__baris = baris

    def setKolom(self, kolom):
        self.__kolom = kolom

    def buatTabel(self, data, columnNames):
        baris = len(data)
        kolom = len(data[0])

        # Calculate the maximum width for each column
        maxWidths = [0] * kolom
        for j in range(kolom):
            for i in range(baris):
                maxWidths[j] = max(maxWidths[j], len(str(data[i][j])))
            # Adjust maximum width based on column names
            maxWidths[j] = max(maxWidths[j], len(columnNames[j]))

        # Print the separator line above the table
        self.printSeparator(maxWidths)

        # Print the title row
        for j in range(kolom):
            print(f"| {columnNames[j]}{' ' * (maxWidths[j] - len(columnNames[j]))} ", end="")
        print("|")

        # Print the separator line below the title row
        self.printSeparator(maxWidths)

        # Print the table data
        for i in range(baris):
            for j in range(kolom):
                if callable(data[i][j]):  # Periksa apakah data adalah metode
                    print(f"| {data[i][j]()}{' ' * (maxWidths[j] - len(str(data[i][j]())))} ", end="")
                else:
                    print(f"| {data[i][j]}{' ' * (maxWidths[j] - len(str(data[i][j])))} ", end="")
            print("|")

            # Print the separator line between rows
            self.printSeparator(maxWidths)

    # Helper method to print the separator line
    def printSeparator(self, maxWidths):
        for width in maxWidths:
            print("+" + "-" * (width + 2), end="")
        print("+")

File: Cpp/Program/Tabel.cpp
/*Saya Nabil Hanif Achmaddiredja mengerjakan Latihan Praktikum 4 dalam 
mata kuliah Desain dan Pemrograman Berorientasi Objek untuk 
keberkahanNya maka saya tidak melakukan kecurangan seperti yang telah dispesifikasikan. Aamiin.*/

#include <iostream>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

class Tabel 
{   
    /*Attribute*/
    private:
        int baris;
        int kolom;

    public:
        /*Constructor*/
        Tabel()
        {

        }
        Tabel(int baris, int kolom){
            this->baris = baris;
            this->kolom = kolom;
        }
        
        /*Getter & Setter*/
        //Getter
        int getBaris() {
            return this->baris;
        }

        int getKolom() {
            return this->kolom;
        }

        //Setter
        void setBaris(int baris) {
            this->baris = baris;
        }

        void setKolom(int kolom) {
            this->kolom = kolom;
        }

        // Method to create and print the table
        void buatTabel(vector<vector<string>> data, vector<string> columnNames) {
            int baris = data.size();
            int kolom = data[0].size();

            // Calculate the maximum width for each column
            vector<int> maxWidths(kolom, 0);
            for (int j = 0; j < kolom; j++) {
                for (int i = 0; i < baris; i++) {
                    maxWidths[j] = max(maxWidths[j], (int)data[i][j].length());
                }
                // Adjust maximum width based on column names
                maxWidths[j] = max(maxWidths[j], (int)columnNames[j].length());
            }

            // Print the separator line above the table
            printSeparator(maxWidths);

            // Print the title row
            for (int j = 0; j < kolom; j++) {
                cout << "| " << setw(maxWidths[j]) << left << columnNames[j] << " ";
            }
            cout << "|" << endl;

            // Print the separator line below the title row
            printSeparator(maxWidths);

            // Print the table data
            for (int i = 0; i < baris; i++) {
                for (int j = 0; j < kolom; j++) {
                    cout << "| " << setw(maxWidths[j]) << left << data[i][j] << " ";
                }
                cout << "|" << endl;

                // Print the separator line between rows
                printSeparator(maxWidths);
            }
        }

        // Helper method to print the separator line
        void printSeparator(vector<int> maxWidths) {
            for (int width : maxWidths) {
                cout << "+" << setfill('-') << setw(width + 2) << "-" << setfill(' ');
            }
            cout << "+" << endl;
        }
    ~Tabel()
    {}
};
